pkg/kp/healthmanager.go

```diff
@@ -0,0 +1,340 @@
+// to write.
+const HealthRetryTime = 5 * time.Second
+
+// consulHealthManager maintains a Consul session for all the local node's health checks,
+// renews it periodically, and refreshes all health checks if it expires.
+type consulHealthManager struct {
+	sessionPub *stream.StringValuePublisher // Publishes the current session
+	done       chan struct{}                // Close this to stop reporting health
+	client     *api.Client                  // Connection to the Consul agent
+	node       string
+	logger     logging.Logger // Logger for health events
+}
+
+// NewHealthManager implements the Store interface. It creates a new HealthManager that
+// uses the Consul Key-Value store to hold app health statues.
+func (c consulStore) newSessionHealthManager(
+	node string,
+	logger logging.Logger,
+) HealthManager {
+	// Create a stream of sessions
+	sessionChan := make(chan string)
+	done := make(chan struct{})
+	// Current time of "Jan 2, 15:04:05" turns into "0102-150405"
+	timeStr := time.Now().Format("0102-150405")
+	go ConsulSessionManager(
+		api.SessionEntry{
+			Name:      fmt.Sprintf("health:%s:%s", node, timeStr),
+			LockDelay: 1 * time.Nanosecond,
+			Behavior:  api.SessionBehaviorDelete,
+			TTL:       TTL.String(),
+		},
+		c.client,
+		sessionChan,
+		done,
+		logger,
+	)
+
+	// Hook the session stream into a publisher
+	pub := stream.NewStringValuePublisher(sessionChan, "")
+
+	return &consulHealthManager{
+		sessionPub: pub,
+		done:       done,
+		client:     c.client,
+		node:       node,
+		logger:     logger,
+	}
+}
+
+// Close cleans up the HealthManager. New health reports will not be published, and
+// existing reports will be removed. Implements the HealthManager interface.
+func (m *consulHealthManager) Close() {
+	close(m.done)
+}
+
+// consulHealthUpdater holds the state needed for the update process to track the current
+// service health and which health it has published.
+type consulHealthUpdater struct {
+	service string           // The service this updater is bound to
+	node    string           // The node this updater is bound to
+	checker chan WatchResult // Stream of health updates from a checker
+}
+
+// NewUpdater creates a new HealthUpdater that can be used to update an app's health status
+// on this node.
+func (m *consulHealthManager) NewUpdater(service string) HealthUpdater {
+	checksStream := make(chan WatchResult)
+	u := &consulHealthUpdater{
+		service: service,
+		node:    m.node,
+		checker: checksStream,
+	}
+	go func() {
+		sub := m.sessionPub.Subscribe(nil)
+		defer sub.Unsubscribe()
+		processHealthUpdater(
+			m.client,
+			checksStream,
+			sub.Chan(),
+			m.logger.SubLogger(logrus.Fields{
+				"service": service,
+				"node":    m.node,
+			}),
+		)
+	}()
+	return u
+}
+
+func (u *consulHealthUpdater) PutHealth(health WatchResult) error {
+	if health.Id != u.service || health.Node != u.node || health.Service != u.service {
```

Comment:
The existing code didn't make any distinction, so my changes didn't either.  In general, I'd say that if a distinction were required, `NewUpdater(service string)` should become `NewUpdater(id, service string)`.  Want that in this stack of commits?